graphix/command.py

```diff
@@ -3,65 +3,92 @@
 from __future__ import annotations
 
 import abc
-import enum
-from typing import TYPE_CHECKING
+import dataclasses
+from abc import ABC
+from enum import Enum
+from typing import Any, Literal
 
 import numpy as np
-from pydantic import BaseModel, ConfigDict
+import typing_extensions
 
 from graphix import clifford
 from graphix.pauli import Pauli, Plane, Sign
 from graphix.states import BasicStates, State
 
-if TYPE_CHECKING:
+if typing_extensions.TYPE_CHECKING:
     from graphix.clifford import Clifford
 
 Node = int
 
 
-class CommandKind(enum.Enum):
+class CommandKind(Enum):
     N = "N"
     M = "M"
     E = "E"
     C = "C"
     X = "X"
     Z = "Z"
-    T = "T"
     S = "S"
 
 
-class Command(BaseModel, abc.ABC):
+class Command(ABC):
     """
     Base command class.
     """
 
-    kind: CommandKind = None
+    @property
+    @abc.abstractmethod
+    def kind(self) -> Any: ...
```

Comment:
I believe `CommandKind` is a more appropriate type for `kind` than `Any`, even though this type is refined in subclasses.

Defining `kind` as a property instead of a field affects performance. On my laptop, running `pytest` takes 20.67 seconds on the `refactor-classes` branch (commit 261acbd), whereas it only takes 17.95 seconds on commit 124e3ab (the latest commit on `master` which this branch is based on). When I replace the `kind` method with a simple field (as in [this commit](https://github.com/thierry-martinez/graphix/commit/7236126c5ef61a4011f69e0c4e55e485e5c83413)), `pytest` takes only 17.10 seconds. While the timings themselves are not entirely consistent from one run to another, the comparison suggests that using a `kind` method is slower than the `pydantic` field version, which is, in turn, slower than the `pydantic`-free field version.

Reply:
```python
"""Validate property."""

import inspect
import typing
from typing import Any, ClassVar, Literal


class CheckedBase:
    """Check the child class."""

    def __init_subclass__(cls, **kwargs: Any) -> None:
        """Check the child class."""
        super().__init_subclass__(**kwargs)
        if not hasattr(cls, "tag"):
            msg = f"{cls.__name__} must have a tag class attribute initialized."
            raise TypeError(msg)
        ann = inspect.get_annotations(cls)
        if "tag" not in ann:
            msg = "Tag attribute must be annotated."
            raise TypeError(msg)
        ann_tag = ann["tag"]
        if typing.get_origin(ann_tag) is not ClassVar:
            msg = "Tag attribute must be a class variable."
            raise TypeError(msg)
        (ann_tag,) = typing.get_args(ann_tag)
        if typing.get_origin(ann_tag) is not Literal:
            msg = "Tag attribute must be a literal."
            raise TypeError(msg)


class Ok(CheckedBase):
    """Valid class."""

    tag: ClassVar[Literal[42]] = 42


class Bad(CheckedBase):
    """Invalid class."""

    spam: int
```

Something like this?

Reply:
Please check d6f0a7c9acd299110087467a0b4de1771ba8e481 .